### PR TITLE
Direct seekable buffer read/write support

### PR DESCRIPTION
**Pull request by sopak**
Sometimes it is useful to access files randomly.
I have implemented java.nio interface for Smbj.
It is buffer oriented.

### CODE REVIEW

**Comment by yin19941005** [comment]
Hi @sopak ,

Nice work! I believe this helps a lot! However, I find a small bug on it. [On this line](https://github.com/hierynomus/smbj/pull/49/files#diff-37771c25f310d7ce8f390c4b585748c2R134), the `smb2Packet.getData()` may null and cause the null pointer exception. I tried to open the file with `AccessMask.GENERIC_WRITE` only (no read access) and try to read the file and this line cause the null pointer exception. I further test for the write file without the write access, the exception is already caught.

I suggest to fix this issue by checking the `smb2Packet.getData()` is null or not first and log the NTSTATUS when it is null for debugging. The code as follow :  
```
if(smb2Packet.getData() != null) {
       System.arraycopy(smb2Packet.getData(), 0, dst, 0, read);
}else {
       logger.info("smb2Packet.getData() is null, NTSTATUS = " + smb2Packet.getHeader().getStatus().toString());
}
```

Do you agree with it?

**Comment by sopak** [comment]
@yin19941005 I agree, but I will fix it by checking access mask for write/read permissions.

> 
        if ((this.accessMask & (AccessMask.GENERIC_READ.getValue() | AccessMask.FILE_READ_EA.getValue())) == 0) {
            throw new SMBApiException(
                NtStatus.STATUS_ACCESS_DENIED,
                NtStatus.STATUS_ACCESS_DENIED.getValue(),
                null,
                "The file is not open for reading"
            );
        }

Do you agree with it?

**Comment by yin19941005** [comment]
Hi @sopak ,

I agree we can handle the case that the granted access is not allow for the read/write operation by your new proposed commit. However, I think we should check for read operation is succeed or not. (For the write operation, it's not matter because the server is responsible and it will return error NTSTATUS for us.)

In my opinion, it seems when any error returned on the read request, the `response.getData()` will be null (and may cause the null pointer exception). It make sense because when error occur, the read operation should be aborted. I just read the [MSDN document on here](https://msdn.microsoft.com/en-us/library/cc246729.aspx), there are actually still others possibility to cause error and crash the program by null pointer exception.

I suggest besides your newly commit to prevent the access right error, we should also check the NTSTATUS is success or not before calling the `System.arraycopy(smb2Packet.getData(), 0, dst, 0, read);` : 
```
if (smb2Packet.getHeader() != null) {
  if (smb2Packet.getHeader().getStatus() == NtStatus.STATUS_SUCCESS) {
   System.arraycopy(smb2Packet.getData(), 0, dst, 0, read);
  }
}
```
What's your opinion about this?

**Comment by sopak** [comment]
@yin19941005  smb2Packet.getHeader() is  never null according (I see Erp never checks for it)
> protected final SMB2Header header = new SMB2Header();

what about  this

```
SMB2ReadResponse smb2Packet = send.get();
if (smb2Packet.getHeader().getStatus() != NtStatus.STATUS_SUCCESS) {
    throw new SMBApiException(smb2Packet.getHeader(), "Read failed for " + this);
}
```

File: src/main/java/com/hierynomus/smbj/share/File.java [hunk 1]
```diff
@@ -96,6 +102,143 @@ public OutputStream getOutputStream(final ProgressListener listener) {
...
+        );
+
+        Future<SMB2ReadResponse> send = session.send(rreq);
+        try {
+            SMB2ReadResponse smb2Packet = send.get();
+            if (smb2Packet.getHeader().getStatus() != NtStatus.STATUS_SUCCESS) {
+                throw new SMBApiException(smb2Packet.getHeader(), "Read failed for " + this);
+            }
+            int read = smb2Packet.getDataLength();
+            if(read>payloadSize){
```
Comment (hierynomus) [inline src/main/java/com/hierynomus/smbj/share/File.java]: formatting is off.

File: src/main/java/com/hierynomus/smbj/share/File.java [hunk 2]
```diff
@@ -96,6 +102,143 @@ public OutputStream getOutputStream(final ProgressListener listener) {
...
+                read=payloadSize;
```
Comment (hierynomus) [inline src/main/java/com/hierynomus/smbj/share/File.java]: here too

File: src/main/java/com/hierynomus/smbj/share/File.java [hunk 3]
```diff
@@ -96,6 +102,143 @@ public OutputStream getOutputStream(final ProgressListener listener) {
...
+    /**
+     * Direct buffer write to the file
+     *
+     * @param src        source buffer to be  written
+     * @param fileOffset offset where to start write
+     * @param lenght     length of src buffer
+     * @return bytes written
+     * @throws IOException
+     */
+    public int write(final byte[] src, final int fileOffset, final int lenght) throws IOException {
```
Comment (hierynomus) [inline src/main/java/com/hierynomus/smbj/share/File.java]: Spelling mistake for `length`

File: src/main/java/com/hierynomus/smbj/share/File.java [hunk 4]
```diff
@@ -96,6 +102,143 @@ public OutputStream getOutputStream(final ProgressListener listener) {
...
+            throw new IOException(e);
+        }
+    }
+
+    /**
+     * Direct buffer write to the file
+     *
+     * @param src        source buffer to be  written
+     * @param fileOffset offset where to start write
+     * @param lenght     length of src buffer
```
Comment (hierynomus) [inline src/main/java/com/hierynomus/smbj/share/File.java]: Spelling mistake for `length`

File: src/main/java/com/hierynomus/smbj/share/File.java [hunk 5]
```diff
@@ -96,6 +102,143 @@ public OutputStream getOutputStream(final ProgressListener listener) {
...
+                NtStatus.STATUS_ACCESS_DENIED.getValue(),
+                null,
+                "The file is not open for writing"
+            );
+        }
+
+        Session session = treeConnect.getSession();
+        NegotiatedProtocol negotiatedProtocol = session.getConnection().getNegotiatedProtocol();
+
+        ByteChunkProvider provider = new ByteChunkProvider()
```
Comment (hierynomus) [inline src/main/java/com/hierynomus/smbj/share/File.java]: Curlies-open go at end of line. Also for a few more methods.

File: src/main/java/com/hierynomus/smbj/share/File.java [hunk 6]
```diff
@@ -96,6 +102,143 @@ public OutputStream getOutputStream(final ProgressListener listener) {
...
+        {
+            int remaining = lenght;
+
+            public ByteChunkProvider withOffset(int offset){
```
Comment (hierynomus) [inline src/main/java/com/hierynomus/smbj/share/File.java]: Space before curly

**Comment by hierynomus** [comment]
Thanks for the PR, few minor comments to polish it :)

**Comment by sopak** [comment]
@hierynomus Hi, may you share your formatting settings for this project? 
I will probably add few more implementations.  And I do not want mess it too much ;).

**Pull request merged**